agent/acs/handler/acs_handler.go

```diff
@@ -327,6 +273,122 @@ func startACSSession(ctx context.Context, client wsclient.ClientServer, timer tt
 	}
 }
 
+// handleACSError handles an error from ACS. It returns true if the session can reconnect
+// to ACS. Else, it returns false. It also applies the appropriate backoff policy for
+// errors which can result in the handler reconnecting to ACS. For an "InactiveInstance"
+// error, it emits an event to the deregister instance event stream.
+func (s *Session) handleACSError(acsError error, connectToACS chan<- struct{}) bool {
+	if acsError == nil || acsError == io.EOF {
+		s.backoff.Reset()
+	} else if strings.HasPrefix(acsError.Error(), "InactiveInstanceException:") {
+		seelog.Debug("Container instance is deregistered, notifying listeners")
+		err := s.deregisterInstanceEventStream.WriteToEventStream(struct{}{})
+		if err != nil {
+			seelog.Debugf("Failed to write to deregister container instance event stream, err: %v", err)
+		}
+		return false
+	} else {
+		seelog.Infof("Error from acs; backing off, err: %v", acsError)
+		s.time().Sleep(s.backoff.Duration())
+	}
+
+	// Received a non-terminal error from ACS. Emit an event to the connectToACS channel
+	go func() {
+		connectToACS <- struct{}{}
+	}()
+	return true
+}
+
+func (s *Session) time() ttime.Time {
+	s.initTime()
+	return s._time
+}
+
+func (s *Session) heartbeatTimeout() time.Duration {
+	s.initTime()
+	return s._heartbeatTimeout
+}
+
+func (s *Session) heartbeatJitter() time.Duration {
+	s.initTime()
+	return s._heartbeatJitter
+}
+
+func (s *Session) initTime() {
```

Comment:
Since you're refactoring, can we get rid of this and just properly inject `time` as well as timeouts?